expression/builtin_cast.go

```diff
@@ -426,25 +426,34 @@ func (c *castAsJSONFunctionClass) getFunction(ctx sessionctx.Context, args []Exp
 
 type builtinCastIntAsIntSig struct {
 	baseBuiltinFunc
```

Comment:
how about defining another struct to wrap `baseBuiltinFunc` and `insertContext`, and use this struct for every `builtinCastXAsYSig`:
```go
type baseBuiltinCastFunc struct {
	baseBuiltinFunc
	inUnionAll bool
}
```

If so, we only need to change the `base` or `bf` when constructing `builtinCastXAsYSig`s, and the code can be cleaner.